src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -198,6 +248,304 @@ ErrorIfIllegalPartitioningInTenantSchema(Oid parentRelationId, Oid partitionRela
+		}
+		else if (IsCitusTableType(relationId, SINGLE_SHARD_DISTRIBUTED))
+		{
+			schemaTables->singleShardTableIdList = lappend_oid(
+				schemaTables->singleShardTableIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, REFERENCE_TABLE))
+		{
+			schemaTables->referenceTableIdList = lappend_oid(
+				schemaTables->referenceTableIdList, relationId);
+		}
+		else if (IsCitusTableType(relationId, CITUS_LOCAL_TABLE))
+		{
+			schemaTables->citusLocalTableIdList = lappend_oid(
+				schemaTables->citusLocalTableIdList, relationId);
+		}
+	}
+	return schemaTables;
+}
+
+
+/*
+ * SchemaHasCitusTableType determines if tables under schema has any Citus
+ * table type given in the list.
+ */
+static bool
+SchemaHasCitusTableType(SchemaTables *schemaTables, List *citusTableTypes)
+{
+	CitusTableType citusTableType = 0;
+	foreach_int(citusTableType, citusTableTypes)
+	{
+		switch (citusTableType)
+		{
+			case HASH_DISTRIBUTED:
+			{
+				if (schemaTables->hashDistributedTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case RANGE_DISTRIBUTED:
+			{
+				if (schemaTables->rangeDistributedTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case APPEND_DISTRIBUTED:
+			{
+				if (schemaTables->appendDistributedTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case SINGLE_SHARD_DISTRIBUTED:
+			{
+				if (schemaTables->singleShardTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case REFERENCE_TABLE:
+			{
+				if (schemaTables->referenceTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			case CITUS_LOCAL_TABLE:
+			{
+				if (schemaTables->citusLocalTableIdList)
+				{
+					return true;
+				}
+				break;
+			}
+
+			default:
+			{
+				ereport(ERROR, (errmsg("unexpected table type when verifying schema "
+									   "tables")));
+			}
+		}
+	}
+	return false;
+}
+
+
+/*
+ * DisallowedTableTypesForDistributeSchema returns Citus tables which are
+ * not allowed inside a schema to be distributed. Currently, we allow
+ * only Citus local and PG local tables.
+ */
+static List *
+DisallowedTableTypesForDistributeSchema(void)
+{
+	List *disallowedCitusTableTypes = NIL;
+	disallowedCitusTableTypes = lappend_int(disallowedCitusTableTypes, HASH_DISTRIBUTED);
+	disallowedCitusTableTypes = lappend_int(disallowedCitusTableTypes,
+											APPEND_DISTRIBUTED);
+	disallowedCitusTableTypes = lappend_int(disallowedCitusTableTypes, RANGE_DISTRIBUTED);
+	disallowedCitusTableTypes = lappend_int(disallowedCitusTableTypes,
+											SINGLE_SHARD_DISTRIBUTED);
+	disallowedCitusTableTypes = lappend_int(disallowedCitusTableTypes, REFERENCE_TABLE);
+	return disallowedCitusTableTypes;
+}
+
+
+/*
+ * DisallowedTableTypesForUndistributeSchema returns Citus tables which are
+ * not allowed inside a schema during the undistribution of the schema. We
+ * expect only single shard tables.
+ */
+static List *
+DisallowedTableTypesForUndistributeSchema(void)
```

Comment:
(doesn't seem to be addressed yet)

Reply:
Yeah I tried it but then reverted since it looks too verbose. i.e. I need the macro at the function declaration, definition, and the place we called the function. What do you think about that?